### DISCUSSION TITLE
Different Kong container images?

### DISCUSSION THREAD

**Discussion by tobiasehlert** [Kong/kong#10515] [Help] [answered]
I'm kind of confused what the difference between following two containers is.

1. https://hub.docker.com/r/kong/kong-gateway
2. https://hub.docker.com/_/kong

Spontaneous it looks to me that the "official image" (bullet 2) is lacking behind in versioning and the kong-gateway (bullet 1) is aligned with the actual releases made here in the kong/kong repo. And also some difference in image size too.
Or is there some other difference that is not fully visible to me?

**Comment by ghost** [comment]
1 maintained by Kong Inc;

2 is `Docker Official Image`, it is a plan launched by Docker. read more: https://docs.docker.com/docker-hub/official_images/

I'm not sure if 2 is also maintained by Kong Inc, but apparently they are under different ownership.

**Reply by tobiasehlert** [reply]
Ah alright!
So basically it doesn't really matter what to choose except for kong/kong-gateway is more updated, but no difference in database scheme or what so ever?

The confusing part was that the https://github.com/Kong/docker-kong repo links to the "docker official" one, so this was the reason for the initial question. I would expect both to be aligned in versions in the first place too hehe

**Reply by ghost** [reply]
> but no difference in database scheme or what so ever?

Not sure, may need to see the original Dockerfile.

**Reply by yogeshjha06** [reply]
I'm evaluating two approaches for deploying Kong Gateway:

1. **Using the Prebuilt Kong Image:**

   ```Dockerfile
   FROM kong/kong-gateway:3.10.0.1-debian
   ```

2. **Building a Custom Image from a Minimal Base:**

   ```Dockerfile
   FROM debian:bullseye-slim

   COPY kong.deb /tmp/kong.deb

   RUN set -ex; \
       apt-get update \
       && apt-get install --yes /tmp/kong.deb \
       && rm -rf /var/lib/apt/lists/* \
       && rm -rf /tmp/kong.deb \
       && chown kong:0 /usr/local/bin/kong \
       && chown -R kong:0 /usr/local/kong \
       && ln -s /usr/local/openresty/luajit/bin/luajit /usr/local/bin/luajit \
       && ln -s /usr/local/openresty/luajit/bin/luajit /usr/local/bin/lua \
       && ln -s /usr/local/openresty/nginx/sbin/nginx /usr/local/bin/nginx \
       && kong version

   COPY docker-entrypoint.sh /docker-entrypoint.sh

   USER kong

   ENTRYPOINT ["/docker-entrypoint.sh"]

   EXPOSE 8000 8443 8001 8444 8002 8445 8003 8446 8004 8447

   STOPSIGNAL SIGQUIT

   HEALTHCHECK --interval=10s --timeout=10s --retries=10 CMD kong health

   CMD ["kong", "docker-start"]
   ```

**Could someone elaborate on the technical differences between these two approaches? Specifically:**

* **Maintenance and Security:** How do updates and security patches differ between the prebuilt image and a custom-built one?

* **Configuration Complexity:** What are the implications for configuration management and setup complexity?

* **Best Practices:** Are there recommended scenarios for choosing one approach over the other?

Any insights or experiences shared would be greatly appreciated!

**Reply by Oyami-Srk** [reply]
Hi @yogeshjha06. The `kong/kong-gateway` is the enterprise version and requires a valid license. For the custom build, I think you can refer to the Dockerfile at https://github.com/Kong/kong/blob/master/build/dockerfiles/deb.Dockerfile.

**Reply by Oyami-Srk** [reply]
For custom builds, the image has not been fully tested, so it may have some issues compared to the official image. If you need to apply future updates, you will have to rebuild the image using the updated packages. On this point, there is no significant difference between prebuilt images and custom-built ones.
In terms of configuration, there is no difference between the two types of images. We recommend using the officially built image, as it has gone through comprehensive testing. We typically release the .deb package and Docker image together. Unless you require a specific base operating system version, we suggest using the official image.

**Reply by yogeshjha06** [reply]
Hey @Oyami-Srk , thanks It really helpful to know about build difference. 
Just to double check `official image` do you mean `custom-build-image` or `prebuild-image` I already on  `prebuild-image`  for Kong 3.4 version but since we are upgrading to 3.10 we wanted to opt best option.
Once again thank you for your detailed clarification.

**Reply by Oyami-Srk** [reply, marked as answer]
@yogeshjha06 For `official image`, I mean the `prebuild-image` published on the Docker Hub.
Enterprise version: https://hub.docker.com/r/kong/kong-gateway
OSS version: https://hub.docker.com/_/kong
